File: 000001-000200/000020.有效的括号/20.js
/**
 * @param {string} s
 * @return {boolean}
 */
// 方法一：用时 76ms
// 执行用时超过了 59%
// 消耗内存超过了 99%
var isValid = function(s) {
  if (s.length % 2 !== 0) {
    return false
  }
  let stack = []
  let lefts = ['(', '{', '[']
  let brackets = {
    ')': '(',
    ']': '[',
    '}': '{'
  }
  let i = 0
  while (i < s.length) {
    if (lefts.includes(s[i])) { // 左边括号
      stack.push(s[i])
      i++
      continue
    }
    let top = stack.pop()
    if (brackets[s[i]] != top) {
      break
    }
    i ++
  }
  return (i == s.length && !stack.length)
};

console.log(isValid('()'))
console.log(isValid('()[]{}'))
console.log(isValid('(]('))
console.log(isValid('([])'))

File: 000001-000200/000020.有效的括号/20.go
package main

import "fmt"

// 方法一：用时 0ms
// 执行用时超过了 94%
// 消耗内存超过了
func isValid(s string) bool {
	if len(s) % 2 == 1 {
		return false
	}
	mmap := map[byte]byte{
		')': '(',
		'}': '{',
		']': '[',
	}
	stack := []byte{}
	for i := 0; i < len(s); i++ {
		if mmap[s[i]] > 0 {
			if len(stack) > 0 && stack[len(stack)-1] == mmap[s[i]] {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		} else {
			stack = append(stack, s[i])
		}
	}
	return len(stack) == 0
}

func main ()  {
	//str := "()"
	//str := "()[]{}"
	//str := "(]"
	//str := "([)]"
	//str := "{[]}"
	str := "){"
	cnt := isValid(str)
	fmt.Println("cnt:", cnt)
}
